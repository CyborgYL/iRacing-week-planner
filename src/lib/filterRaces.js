// @flow

import intersection from 'lodash.intersection';
import type { SeriesRace } from './races';
import type { FilterOptions } from '../reducers/settings';

type FilterArguments = {
  races: Array<SeriesRace>,
  filters: FilterOptions,
  ownedTracks: Array<number>,
  ownedCars: Array<number>,
  favouriteSeries: Array<number>,
  favouriteCars: Array<number>,
  favouriteTracks: Array<number>,
};

export default function filterRaces({
  races, filters, ownedTracks, ownedCars, favouriteSeries, favouriteCars, favouriteTracks,
}: FilterArguments) {
  let filteredRaces: Array<SeriesRace> = [...races];

  filteredRaces = filteredRaces.filter((race) => filters.type.indexOf(race.type) !== -1);

  filteredRaces = filteredRaces.filter((race) => filters.licence.indexOf(race.licenceClass) !== -1);

  filteredRaces = filteredRaces.filter((race) => filters.fixed.indexOf(race.fixed) !== -1);

  filteredRaces = filteredRaces.filter((race) => filters.official.indexOf(race.official) !== -1);

  if (filters.ownedTracks) {
    filteredRaces = filteredRaces.filter((race) => ownedTracks.indexOf(race.trackId) !== -1);
  }

  if (filters.ownedCars) {
    filteredRaces = filteredRaces.filter((race) => intersection(ownedCars, race.carIds).length !== 0);
  }

  if (filters.favouriteSeries) {
    filteredRaces = filteredRaces.filter((race) => favouriteSeries.indexOf(race.seriesId) !== -1);
  }

  if (filters.favouriteCarsOnly) {
    filteredRaces = filteredRaces.filter((race) => intersection(favouriteCars, race.carIds).length !== 0);
  }

  if (filters.favouriteTracksOnly) {
    filteredRaces = filteredRaces.filter((race) => favouriteTracks.indexOf(race.trackId) !== -1);
  }

  return filteredRaces;
}
