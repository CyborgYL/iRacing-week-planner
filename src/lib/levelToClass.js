// @flow

const levels: Array<string> = ['R', 'D', 'C', 'B', 'A', 'P', 'P'];

function effectiveLevelToClass(level: number): string {
  const index = Math.floor((level) / 4);

  return levels[index];
}

function realLevelToClass(level: number): string {
  const index = Math.floor((level - 1) / 4);

  return levels[index];
}

export default function levelToClass(level: number, effective: boolean = false): string {
  return effective ? effectiveLevelToClass(level) : realLevelToClass(level);
}

export function levelToClassNumber(level: number): number {
  return Math.floor((level) / 4);
}
