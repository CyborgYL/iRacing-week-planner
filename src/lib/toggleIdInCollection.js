export default function toggleIdInCollection(collection, id, newState) {
  const newCollection = [...collection];
  const index = newCollection.indexOf(id);

  if (index === -1 && newState) {
    newCollection.push(id);
  }
  if (index !== -1 && newState === false) {
    newCollection.splice(index, 1);
  }
  return newCollection;
}
