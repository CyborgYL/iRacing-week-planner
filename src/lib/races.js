// @flow

import moment, { duration } from 'moment';

import season from '../data/season.json';
import levelToClass, { levelToClassNumber } from './levelToClass';
import raceTimesArray from '../data/raceTimes';

const raceTimesById = raceTimesArray.reduce((races, race) => ({ ...races, [race.seriesId]: race }), {});

function getStartOfWeek(date) {
  return moment(date).subtract(1, 'days').startOf('isoWeek').add(1, 'days');
}

export type TimeableRace = {
  seriesId: number,
  everyTime?: moment.Duration,
  offset?: moment.Duration,
  setTimes?: Array<moment.Duration>,
}

function getNextRaceFromRecur(now, everyTime, offset) {
  const startOfWeek = getStartOfWeek(now);

  const nextDate = startOfWeek.clone().add(offset);
  const endDate = startOfWeek.clone().add({ weeks: 1, days: 1 });

  while (nextDate.isBefore(now) && nextDate.isBefore(endDate)) {
    nextDate.add(everyTime);
  }

  if (nextDate.isBefore(now)) {
    return null;
  }

  return nextDate;
}

function getNextRaceSetTimes(now, setTimes) {
  const startOfWeek = getStartOfWeek(now);

  const nextTime = setTimes.find((time) => startOfWeek.clone().add(time).isAfter(now));

  if (nextTime) {
    return startOfWeek.clone().add(nextTime);
  }

  const firstNextWeek = startOfWeek.clone().add(1, 'weeks').add(setTimes[0]);

  if (firstNextWeek.isAfter(now)) {
    return firstNextWeek;
  }

  return null;
}

export function getNextRace(date: moment.Moment, race: TimeableRace): ?moment.Moment {
  if (race.everyTime) {
    return getNextRaceFromRecur(date, race.everyTime, race.offset);
  }

  if (race.setTimes) {
    return getNextRaceSetTimes(date, race.setTimes);
  }

  return null;
}

const getType = (catId) => {
  const categories = {
    1: 'Oval',
    2: 'Road',
    3: 'Dirt',
    4: 'RX',
  };

  return categories[catId];
};

export default season.reduce((carry, series) => {
  const raceTimes = raceTimesById[series.seriesid] || {};

  const seriesName = series.seriesname;
  const seriesStart = moment(series.start, 'x').utc().startOf('day');

  if (raceTimes.weekStartOffset) {
    seriesStart.add(raceTimes.weekStartOffset);
  }

  const seriesEnd = moment(series.end, 'x').utc().startOf('isoWeek').add({ days: 1 });

  if (raceTimes.weekEndOffset) {
    seriesEnd.add(raceTimes.weekEndOffset);
  }

  const offWeeks = raceTimes.offWeeks || [];
  const raceWeekLength = Math.round(moment(seriesEnd).diff(seriesStart) / (series.tracks.length + offWeeks.length));

  const allRaceWeeks = series.tracks.map((track) => track.raceweek)
    .concat(offWeeks.map((offWeek) => offWeek - 1));

  allRaceWeeks.sort((a, b) => a - b);

  return carry.concat(series.tracks.map((track) => {
    const realRaceWeek = allRaceWeeks.indexOf(track.raceweek);

    const startTime = moment(seriesStart).add(raceWeekLength * realRaceWeek, 'ms').startOf('day').utc();
    const weekLength = duration(raceWeekLength);

    const type = getType(series.catid);

    return {
      series: seriesName,
      seriesId: series.seriesid,
      track: track.name,
      trackId: track.pkgid,
      week: track.raceweek,
      startTime,
      weekLength,
      endTime: moment(startTime).add(weekLength),
      official: series.isOfficial,
      licenceLevel: series.minlicenselevel,
      licenceClassNumber: levelToClassNumber(series.minlicenselevel),
      licenceClass: levelToClass(series.minlicenselevel, true),
      type,
      fixed: series.isFixedSetup,
      carClasses: series.carclasses.map(({ shortname }) => shortname),
      carIds: series.cars.map(({ sku }) => sku),
      seriesStart,
      seriesEnd,
      seasonId: series.seasonid,
      everyTime: raceTimes.everyTime,
      offset: raceTimes.offset,
      setTimes: raceTimes.setTimes,
    };
  }));
}, []);
