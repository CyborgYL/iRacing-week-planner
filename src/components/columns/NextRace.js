// @flow

import * as React from 'react';
import moment from 'moment';
import { useTranslation } from 'react-i18next';
import type { TimeableRace } from '../../lib/races';
import { getNextRace } from '../../lib/races';

type Props = {
  race: TimeableRace,
};

export default function NextRace({ race }: Props) {
  const { t } = useTranslation();

  const now = moment().utc();

  // Only update at most every minute
  const nowMemoizer = now.format('YYYY-MM-DD HH-mm');

  const nextTime = React.useMemo(() => getNextRace(now, race), [race, nowMemoizer]);

  if (nextTime === null) {
    return <td>{t('No time data')}</td>;
  }

  const date = moment(nextTime).local().toDate();

  return (
    <td>
      <div>
        {t('{{date, ddd k:mm}}', { date })}
      </div>
    </td>
  );
}
