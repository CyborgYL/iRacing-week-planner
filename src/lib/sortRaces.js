import columns from '../data/availableColumns';

const columnsById = columns.reduce((columnObject, column) => ({ ...columnObject, [column.id]: column }), {});

export default (rule, unordered) => {
  const races = [...unordered];

  races.sort(columnsById[rule.key].sort.bind(null, rule.order));

  return races;
};
