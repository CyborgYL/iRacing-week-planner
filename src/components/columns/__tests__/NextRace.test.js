import { describe, test, expect } from '@jest/globals';
import MockDate from 'mockdate';
import moment from 'moment';
import React from 'react';
import { shallow } from 'enzyme';
import { getNextRace } from '../../../lib/races';

import NextRace from '../NextRace';

jest.mock(
  '../../../lib/races',
  () => ({
    getNextRace: jest.fn(),
  }),
);

describe('components/columns/NextRace', () => {
  test('renders correctly', () => {
    MockDate.set('2020-09-23T13:30:00.000Z');
    getNextRace.mockReturnValue(moment('2020-09-23T13:45:00.000Z'));
    const component = shallow(<NextRace race={{}} />);

    expect(component).toMatchSnapshot();
    MockDate.reset();
  });

  test('renders no time data', () => {
    MockDate.set('2020-09-23T13:30:00.000Z');
    getNextRace.mockReturnValue(null);
    const component = shallow(<NextRace race={{}} />);

    expect(component).toMatchSnapshot();
    MockDate.reset();
  });
});
