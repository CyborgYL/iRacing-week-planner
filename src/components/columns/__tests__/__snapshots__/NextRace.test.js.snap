// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/columns/NextRace renders correctly 1`] = `
<td>
  <div>
    Wed 13:45
  </div>
</td>
`;

exports[`components/columns/NextRace renders no time data 1`] = `
<td>
  No time data
</td>
`;
